src/compile-middleware.ts

```diff
@@ -84,56 +104,135 @@ export function babelCompile(forceCompile: boolean): RequestHandler {
+      } else if (javaScriptMimeTypes.includes(contentType)) {
+        transformed = compileScript(body, options);
+      } else {
+        transformed = body;
       }
-      babelCompileCache.set(source, body);
-
-      return body;
+      babelCompileCache.set(body, transformed);
+      return transformed;
     },
   });
 }
 
-function compileHtml(source: string, location: string): string {
+function compileHtml(
+    source: string, location: string, options: compileOptions): string {
   const document = parse5.parse(source);
-  const scriptTags = dom5.queryAll(document, isInlineJavaScript);
-  for (const scriptTag of scriptTags) {
-    try {
-      const script = dom5.getTextContent(scriptTag);
-      const compiledScriptResult = compileScript(script);
-      dom5.setTextContent(scriptTag, compiledScriptResult);
-    } catch (e) {
-      // By not setting textContent we keep the original script, which
-      // might work. We may want to fail the request so a better error
-      // shows up in the network panel of dev tools. If this is the main
-      // page we could also render a message in the browser.
-      console.warn(`Error compiling script in ${location}: ${e.message}`);
+  let injectedRequireJS = false;
+
+  for (const scriptTag of dom5.queryAll(document, isJsScriptNode)) {
+    // Is this a module script we should transform?
+    const transformingModule = options.transformModules &&
+        dom5.getAttribute(scriptTag, 'type') === 'module';
+
+    if (transformingModule && !injectedRequireJS) {
+      // We need RequireJS to load the AMD modules we are declaring. Inject the
+      // dependency as late as possible (right before the first module is
+      // declared) because some of our legacy non-module dependencies,
+      // typically loaded in <head>, behave differently when window.require is
+      // present.
+      dom5.insertBefore(
+          scriptTag.parentNode,
+          scriptTag,
+          parse5.parseFragment(
+              '<script src="/components/requirejs/require.js"></script>\n'));
+      injectedRequireJS = true;
+    }
+
+    const src = dom5.getAttribute(scriptTag, 'src');
+    const isInline = !src;
+
+    if (transformingModule && !isInline) {
+      // Transform an external module script into a `require` for that module,
+      // to be executed immediately.
+      dom5.replace(
+          scriptTag,
+          parse5.parseFragment(`<script>require(["${src}"]);</script>\n`));
+
+    } else if (isInline) {
+      const plugins = [];
+      if (options.transformES2015) {
+        plugins.push(...es2015Plugins);
+      }
+      if (transformingModule) {
+        plugins.push(...modulesPlugins);
+      }
+
+      const js = dom5.getTextContent(scriptTag);
+      let compiled;
+      try {
+        compiled = babelCore.transform(js, {plugins}).code;
+      } catch (e) {
+        // By not setting textContent we keep the original script, which might
```

Comment:
Cleaned up the comment a bit and added a TODO. (Note this behavior was the same before this PR).